packages/sdk/src/cosmwasmclient.spec.ts

```diff
@@ -0,0 +1,222 @@
+      expect(client).toBeTruthy();
+    });
+  });
+
+  describe("chainId", () => {
+    it("works", async () => {
+      pendingWithoutCosmos();
+      const client = CosmWasmClient.makeReadOnly(httpUrl);
+      expect(await client.chainId()).toEqual("testing");
+    });
+  });
+
+  describe("getNonce", () => {
+    it("works", async () => {
+      pendingWithoutCosmos();
+      const client = CosmWasmClient.makeReadOnly(httpUrl);
+      expect(await client.getNonce(unusedAccount.address)).toEqual({
+        accountNumber: 5,
+        sequence: 0,
+      });
+    });
+  });
+
+  describe("postTx", () => {
+    it("works", async () => {
+      pendingWithoutCosmos();
+      const pen = await Secp256k1Pen.fromMnemonic(faucet.mnemonic);
+      const client = CosmWasmClient.makeReadOnly(httpUrl);
+
+      const memo = "My first contract on chain";
+      const sendMsg: MsgSend = {
+        type: "cosmos-sdk/MsgSend",
+        value: {
+          // eslint-disable-next-line @typescript-eslint/camelcase
+          from_address: faucet.address,
+          // eslint-disable-next-line @typescript-eslint/camelcase
+          to_address: makeRandomAddress(),
+          amount: [
+            {
+              denom: "ucosm",
+              amount: "1234567",
+            },
+          ],
+        },
+      };
+
+      const fee: StdFee = {
+        amount: [
+          {
+            amount: "5000",
+            denom: "ucosm",
+          },
+        ],
+        gas: "890000",
+      };
+
+      const chainId = await client.chainId();
+      const { accountNumber, sequence } = await client.getNonce(faucet.address);
+      const signBytes = makeSignBytes([sendMsg], fee, chainId, memo, accountNumber, sequence);
+      const signature = encodeSecp256k1Signature(pen.pubkey, await pen.createSignature(signBytes));
+      const signedTx = {
+        msg: [sendMsg],
+        fee: fee,
+        memo: memo,
+        signatures: [signature],
+      };
+      const { logs } = await client.postTx(marshalTx(signedTx));
+      const amountAttr = findAttribute(logs, "transfer", "amount");
+      expect(amountAttr.value).toEqual("1234567ucosm");
+    });
+  });
+
+  describe("upload", () => {
+    it("works", async () => {
+      pendingWithoutCosmos();
+      const pen = await Secp256k1Pen.fromMnemonic(faucet.mnemonic);
+      const client = CosmWasmClient.makeWritable(httpUrl, faucet.address, async signBytes => {
+        return encodeSecp256k1Signature(pen.pubkey, await pen.createSignature(signBytes));
+      });
+      const codeId = await client.upload(getRandomizedHackatom());
+      expect(codeId).toBeGreaterThanOrEqual(1);
+    });
+  });
+
+  describe("instantiate", () => {
+    it("works with transfer amount", async () => {
+      pendingWithoutCosmos();
+      const pen = await Secp256k1Pen.fromMnemonic(faucet.mnemonic);
+      const client = CosmWasmClient.makeWritable(httpUrl, faucet.address, async signBytes => {
```

Comment:
This is nice, but can we extend Pen to provide this closure?

eg. `makeWritable(httpUrl, facuet.address, pen.signer())` where

`signer(): SigningCallback`?

Or if that violates encapsulation, a helper like 
`CosmWasmClient.makeWritableFromPen(httpUrl, pen)`
and it can use `pen.pubkey` to generate the signing address as well

Reply:
Both ways make sense to me. I'll think about how to implement this feature without creating unwanted dependencies.